[WPE] `TestWebKitWebView` `/webkit/WebKitWebView/fullscreen` is a timeout

obyknovenius opened the pull request:
#### 59f29abe483ee8b33b50f5f476f88f67ef5836d8
<pre>
[WPE] `TestWebKitWebView` `/webkit/WebKitWebView/fullscreen` is a timeout
<a href="https://bugs.webkit.org/show_bug.cgi?id=258850">https://bugs.webkit.org/show_bug.cgi?id=258850</a>

Reviewed by Michael Catanzaro.

On WPE `WKWPE::View` creates a web page with its own page client.

But when using GLib API we must somehow notify the wrapper `WebKitWebView`
about entering and exiting fullscreen mode.

The solution is to extend `WKWPE::API::Client` with two new methods:
- `enterFullScreen()`
- `exitFullScreen()`

`WebKitWebViewClient`, which implements `WKWPE::API::Client` and has
a reference to `WebKitWebView` wrapper, can react on these events and
call appropriate handlers.

* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
(WebKitWebViewClient::enterFullScreen):
(WebKitWebViewClient::exitFullScreen):
* Source/WebKit/UIProcess/API/wpe/APIViewClient.h:
(API::ViewClient::enterFullScreen):
(API::ViewClient::exitFullScreen):
* Source/WebKit/UIProcess/API/wpe/PageClientImpl.cpp:
(WebKit::PageClientImpl::enterFullScreen):
(WebKit::PageClientImpl::exitFullScreen):
* Source/WebKit/UIProcess/API/wpe/WPEView.cpp:
(WKWPE::View::enterFullScreen):
(WKWPE::View::exitFullScreen):
* Source/WebKit/UIProcess/API/wpe/WPEView.h:
* Source/WebKit/UIProcess/API/wpe/WebKitWebViewClient.h:
* Tools/TestWebKitAPI/glib/TestExpectations.json:

Canonical link: <a href="https://commits.webkit.org/265799@main">https://commits.webkit.org/265799@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6d2a60c44b889b06b1cd23a86ae26b46e8d7b104

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/11943 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/12142 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/12519 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/13588 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/11369 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/11962 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/14531 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/12123 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/14229 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/12107 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/12920 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/10101 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/14011 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/10212 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/10837 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/17967 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/11290 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/10996 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/14166 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/11412 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/9439 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/10590 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2879 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/14872 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/11269 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WebKit/UIProcess/API/wpe/WPEView.cpp:
```diff
@@ -363,6 +363,18 @@ WebKitWebResourceLoadManager* View::webResourceLoadManager()
+{
+    m_client->exitFullScreen(*this);
+}
+#endif
```
I don't think these are really needed. Wouldn't it be better to call `m_client->enter/exitFullScreen` from `View::setFullScreen`? Otherwise, it's really unclear why you have to separately call `setFullScreen` and `enterFullScreen` or `exitFullScreen` on the same WPEView.

obyknovenius replied:
In my original implementation, I used `View::setFullScreen()` only.  But then I noticed that for every `API::ViewClient` method there is a corresponding method in `WKWPE::View`:
https://github.com/WebKit/WebKit/blob/13fa2acdd4842002cebe0ba83ca272cb4213f391/Source/WebKit/UIProcess/API/wpe/WPEView.h#L76-L82
I thought that it might be a good idea to keep this pattern. But I'm not sure. If you still think we don't need these separate API calls I can get rid of them. So, what would you say?

mcatanzaro replied:
If you want them to match, then you can replace `WPEView::setFullScreen` with `enterFullScreen` and `exitFullScreen`. But you shouldn't have them alongside `setFullScreen` as having two different functions that semantically sound identical but don't do the same thing is too confusing.

obyknovenius replied:
I decided not to change the existing API and do what you originally suggested.

mcatanzaro reviewed (changes requested):
.

obyknovenius committed:
[WPE] `TestWebKitWebView` `/webkit/WebKitWebView/fullscreen` is a timeout
https://bugs.webkit.org/show_bug.cgi?id=258850

Reviewed by Michael Catanzaro.

On WPE `WKWPE::View` creates a web page with its own page client.

But when using GLib API we must somehow notify the wrapper `WebKitWebView`
about entering and exiting fullscreen mode.

The solution is to extend `WKWPE::API::Client` with two new methods:
- `enterFullScreen()`
- `exitFullScreen()`

`WebKitWebViewClient`, which implements `WKWPE::API::Client` and has
a reference to `WebKitWebView` wrapper, can react on these events and
call appropriate handlers.

* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
(WebKitWebViewClient::enterFullScreen):
(WebKitWebViewClient::exitFullScreen):
* Source/WebKit/UIProcess/API/wpe/APIViewClient.h:
(API::ViewClient::enterFullScreen):
(API::ViewClient::exitFullScreen):
* Source/WebKit/UIProcess/API/wpe/PageClientImpl.cpp:
(WebKit::PageClientImpl::enterFullScreen):
(WebKit::PageClientImpl::exitFullScreen):
* Source/WebKit/UIProcess/API/wpe/WPEView.cpp:
(WKWPE::View::enterFullScreen):
(WKWPE::View::exitFullScreen):
* Source/WebKit/UIProcess/API/wpe/WPEView.h:
* Source/WebKit/UIProcess/API/wpe/WebKitWebViewClient.h:
* Tools/TestWebKitAPI/glib/TestExpectations.json:

Canonical link: https://commits.webkit.org/265799@main
